### PR TITLE
Source Notion: Add suggested Streams and Optional Start Date

### PR DESCRIPTION
**Pull request by ChristoGrab**
## What
Wrapping up Certification checklist tasks for source Notion.

## How
- Added suggested streams based on usage (all four streams meet the threshold)
- Start Date is no longer a required input, and defaults to 2 years ago if not set by the user
- Switched the stream used for verifying connection from Users to Pages. Pages stream is the most widely used, and reading Page content is permitted by default when setting up credentials (whereas Users permissions must be added manually). This way a customer who just wants to sync their Pages won't be forced to set extra permissions to validate connection.

## User Impact
No impact on existing connections.

### CODE REVIEW

**Comment by airbyte-oss-build-runner** [comment]
##  source-notion test report (commit [`64e906b70f`](https://github.com/airbytehq/airbyte/pull/30780/commits/64e906b70ffe9d0e74bd49ac48500387ea0b9c50)) - ❌

⏲️  Total pipeline duration: 04mn28s 

| Step                                                                 | Result   |
|:---------------------------------------------------------------------|:---------|
| Connector package install                                            | ✅        |
| Build source-notion docker image for platform linux/x86_64           | ✅        |
| Unit tests                                                           | ✅        |
| Acceptance tests                                                     | ❌        |
| Code format checks                                                   | ✅        |
| Validate airbyte-integrations/connectors/source-notion/metadata.yaml | ✅        |
| Connector version semver check                                       | ✅        |
| Connector version increment check                                    | ✅        |
| QA checks                                                            | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/christo_notion-suggested-streams/1695835385/64e906b70ffe9d0e74bd49ac48500387ea0b9c50/source-notion/1.2.0/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:64e906b70ffe9d0e74bd49ac48500387ea0b9c50)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-notion test
```

File: airbyte-integrations/connectors/source-notion/source_notion/source.py [hunk 1]
```diff
@@ -33,10 +34,21 @@ def get_access_token(self):
 
 
 class SourceNotion(AbstractSource):
-    def check_connection(self, logger: AirbyteLogger, config: Mapping[str, Any]) -> Tuple[bool, any]:
+    @staticmethod
+    def set_start_date(config: Mapping[str, Any]):
```
Comment (maxi297) [inline airbyte-integrations/connectors/source-notion/source_notion/source.py:38]: Would this have been safer to implement in `NotionStream.__init__`? I fear that every time we are instantiating a stream now, we need to think about calling `set_start_date` before. If we forget, the behavior might not be the one we expect

**Review by maxi297 — COMMENTED** [review]
I would like to make sure those things are understood before we decide to go with this solution

File: airbyte-integrations/connectors/source-notion/source_notion/source.py [hunk 2]
```diff
@@ -33,10 +34,21 @@ def get_access_token(self):
...
+        """
+        start_date = config.get("start_date")
+        if not start_date:
+            two_years_ago = pendulum.now().subtract(years=2).in_timezone("UTC").format("YYYY-MM-DDTHH:mm:ss.SSS[Z]")
+            config["start_date"] = two_years_ago
+
+    def check_connection(self, logger: logging.Logger, config: Mapping[str, Any]) -> Tuple[bool, any]:
         try:
+            self.set_start_date(config)
             authenticator = NotionAuthenticator(config).get_access_token()
```
Comment (artem1205) [inline airbyte-integrations/connectors/source-notion/source_notion/source.py:50]:
```suggestion
            authenticator = self._get_authenticator(config)
```

File: airbyte-integrations/connectors/source-notion/source_notion/source.py [hunk 3]
```diff
@@ -33,10 +34,21 @@ def get_access_token(self):
...
+    @staticmethod
+    def set_start_date(config: Mapping[str, Any]):
+        """
+        Sets the start date to two years ago if it is not already set by the user.
+        """
+        start_date = config.get("start_date")
+        if not start_date:
+            two_years_ago = pendulum.now().subtract(years=2).in_timezone("UTC").format("YYYY-MM-DDTHH:mm:ss.SSS[Z]")
+            config["start_date"] = two_years_ago
+
```
Comment (artem1205) [inline airbyte-integrations/connectors/source-notion/source_notion/source.py:46]: replace extra class `NotionAuthenticator` with method:  
```suggestion
def _get_authenticator(self, config: Mapping[str, Any]):
    if "access_token" in self.config:
        token = self.config.get("access_token")
    elif:  # TODO: add else cases
        ...
    return TokenAuthenticator(token)
```

File: airbyte-integrations/connectors/source-notion/source_notion/spec.json [hunk 4]
```diff
@@ -4,19 +4,18 @@
...
     "type": "object",
-    "required": ["start_date"],
     "properties": {
       "start_date": {
         "title": "Start Date",
-        "description": "UTC date and time in the format 2017-01-25T00:00:00.000Z. Any data before this date will not be replicated.",
+        "description": "UTC date and time in the format YYYY-MM-DDTHH:MM:SS.000Z. Any data before this date will not be replicated. If left blank, the start date will be set to 2 years before the present date.",
         "pattern": "^[0-9]{4}-[0-9]{2}-[0-9]{2}T[0-9]{2}:[0-9]{2}:[0-9]{2}.[0-9]{3}Z$",
         "examples": ["2020-11-16T00:00:00.000Z"],
         "type": "string",
```
Comment (artem1205) [inline airbyte-integrations/connectors/source-notion/source_notion/spec.json:13]: please add `"pattern_descriptor"`

**Review by artem1205 — CHANGES_REQUESTED** [review]

File: airbyte-integrations/connectors/source-notion/source_notion/source.py (hunk 1 above)
Comment (ChristoGrab) [reply airbyte-integrations/connectors/source-notion/source_notion/source.py]: That's a valid concern! To guarantee the `start_date` is initiated properly in each stream I've moved the logic to `NotionStream.__init__` as suggested. If no `start_date` is found in config, the stream will set it using the two year logic and update the value in config for the next streams to use as well.

**Comment by airbyte-oss-build-runner** [comment]
##  source-notion test report (commit [`2216991172`](https://github.com/airbytehq/airbyte/pull/30780/commits/22169911721252de33eff158165616ca2b699e02)) - ❌

⏲️  Total pipeline duration: 03mn55s 

| Step                                                                 | Result   |
|:---------------------------------------------------------------------|:---------|
| Connector package install                                            | ✅        |
| Build source-notion docker image for platform linux/x86_64           | ✅        |
| Unit tests                                                           | ✅        |
| Acceptance tests                                                     | ❌        |
| Code format checks                                                   | ❌        |
| Validate airbyte-integrations/connectors/source-notion/metadata.yaml | ✅        |
| Connector version semver check                                       | ✅        |
| Connector version increment check                                    | ✅        |
| QA checks                                                            | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/christo_notion-suggested-streams/1695875863/22169911721252de33eff158165616ca2b699e02/source-notion/1.2.0/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:22169911721252de33eff158165616ca2b699e02)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-notion test
```

File: airbyte-integrations/connectors/source-notion/source_notion/source.py [hunk 5]
```diff
@@ -33,10 +34,21 @@ def get_access_token(self):
...
+    def check_connection(self, logger: logging.Logger, config: Mapping[str, Any]) -> Tuple[bool, any]:
         try:
+            self.set_start_date(config)
             authenticator = NotionAuthenticator(config).get_access_token()
-            stream = Users(authenticator=authenticator, config=config)
+            stream = Pages(authenticator=authenticator, config=config)
```
[the remark this answers is not in the document]
Comment (ChristoGrab) [reply airbyte-integrations/connectors/source-notion/source_notion/source.py:31]: If I understand the concern, you're asking about the repercussions of this change on an existing customer that has enabled permissions for the `Users` stream but not `Pages`? If so, based on the way permissions are set when creating a Notion integration, the odds that this is the case are extremely low.

Every Notion credential has `read content` permissions enabled by default (this covers all existing streams other than `Users`). It's impossible to only provide `User` permissions, as at least one other permission has to be granted.

permissions

Essentially, a customer would have to go out of their way to replace the `read content` permission with something like `Insert content` and deny the integration access to any stream but `Users`. Metabase reports also indicate the `Pages` stream is the most widely-used, so I feel confident all existing connections have the permissions to handle this change gracefully.

Let me know if I misunderstood the question!

File: airbyte-integrations/connectors/source-notion/source_notion/source.py (hunk 3 above)
Comment (ChristoGrab) [reply airbyte-integrations/connectors/source-notion/source_notion/source.py:27]: Hi Artem, thanks for the feedback, I've made the changes to the authentication handling as requested.

File: airbyte-integrations/connectors/source-notion/source_notion/spec.json (hunk 4 above)
Comment (ChristoGrab) [reply airbyte-integrations/connectors/source-notion/source_notion/spec.json:14]: Added 👍

**Comment by maxi297** [comment]
My concerns have been addressed. I'll wait for @artem1205 to confirm it's fine on his side too. Thanks @ChristoGrab !

File: airbyte-integrations/connectors/source-notion/source_notion/source.py (hunk 3 above)
Comment (artem1205) [reply airbyte-integrations/connectors/source-notion/source_notion/source.py:27]: LGTM, please check `# support the old config`.
Will approve after that.

**Comment by airbyte-oss-build-runner** [comment]
##  source-notion test report (commit [`368d5760ad`](https://github.com/airbytehq/airbyte/pull/30780/commits/368d5760adfa4888c9e959d885eae37a3fca3895)) - ❌

⏲️  Total pipeline duration: 03mn27s 

| Step                                                       | Result   |
|:-----------------------------------------------------------|:---------|
| Connector package install                                  | ✅        |
| Build source-notion docker image for platform linux/x86_64 | ✅        |
| Unit tests                                                 | ✅        |
| Acceptance tests                                           | ❌        |
| Code format checks                                         | ❌        |
| Validate metadata for source-notion                        | ✅        |
| Connector version semver check                             | ✅        |
| Connector version increment check                          | ✅        |
| QA checks                                                  | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/christo_notion-suggested-streams/1696261766/368d5760adfa4888c9e959d885eae37a3fca3895/source-notion/1.2.0/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:368d5760adfa4888c9e959d885eae37a3fca3895)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-notion test
```

File: airbyte-integrations/connectors/source-notion/source_notion/source.py [hunk 6]
```diff
@@ -15,30 +16,26 @@
...
-        # support the old config
-        if "access_token" in self.config:
-            return TokenAuthenticator(self.config.get("access_token"))
+        if credentials and token:
+            return TokenAuthenticator(token)
 
+        # The original implementation did not support OAuth, and therefore had no "credentials" key.
+        # We can maintain backwards compatibility for OG connections by checking for the deprecated "access_token" key, just in case.
+        if config.get("access_token"):
+            return TokenAuthenticator(config["access_token"])
```
Comment (artem1205) [inline airbyte-integrations/connectors/source-notion/source_notion/source.py:31]: nit, 
```suggestion
        return TokenAuthenticator(config.get"access_token", token))
```

LGTM!

**Review by artem1205 — APPROVED** [review]

**Comment by airbyte-oss-build-runner** [comment]
##  source-notion test report (commit [`87ecc0a268`](https://github.com/airbytehq/airbyte/pull/30780/commits/87ecc0a26875a84329165ca2610d26323e09ad21)) - ✅

⏲️  Total pipeline duration: 03mn41s 

| Step                                                          | Result   |
|:--------------------------------------------------------------|:---------|
| Connector package install                                     | ✅        |
| Build source-notion docker image for platform(s) linux/x86_64 | ✅        |
| Unit tests                                                    | ✅        |
| Acceptance tests                                              | ✅        |
| Code format checks                                            | ✅        |
| Validate metadata for source-notion                           | ✅        |
| Connector version semver check                                | ✅        |
| Connector version increment check                             | ✅        |
| QA checks                                                     | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/christo_notion-suggested-streams/1696858826/87ecc0a26875a84329165ca2610d26323e09ad21/source-notion/1.2.2/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:87ecc0a26875a84329165ca2610d26323e09ad21)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-notion test
```

**Pull request merged**